nikola/utils.py

```diff
@@ -2093,3 +2094,171 @@ def rss_writer(rss_obj, output_path):
         if isinstance(data, bytes_str):
             data = data.decode('utf-8')
         rss_file.write(data)
+
+
+def sort_classifications(taxonomy, classifications, lang):
+    """Sort the given list of classifications of the given taxonomy and language.
+
+    ``taxonomy`` must be a ``Taxonomy`` plugin.
+    ``classifications`` must be an iterable collection of
+    classification strings for that taxonomy.
+    ``lang`` is the language the classifications are for.
+
+    The result will be returned as a sorted list. Sorting will
+    happen according to the way the complete classification
+    hierarchy for the taxonomy is sorted.
+    """
+    if taxonomy.has_hierarchy:
+        # To sort a hierarchy of classifications correctly, we first
+        # build a tree out of them (and mark for each node whether it
+        # appears in the list), then sort the tree node-wise, and finally
+        # collapse the tree into a list of recombined classifications.
+
+        # Step 1: build hierarchy. Here, each node consists of a boolean
+        # flag (node appears in list) and a dictionary mapping path elements
+        # to nodes.
+        root = [False, {}]
+        for classification in classifications:
+            node = root
+            for elt in taxonomy.extract_hierarchy(classification):
+                if elt not in node[1]:
+                    node[1][elt] = [False, {}]
+                node = node[1][elt]
+            node[0] = True
+        # Step 2: sort hierarchy. The result for a node is a pair
+        # (flag, subnodes), where subnodes is a list of pairs (name, subnode).
+
+        def sort_node(node, level=0):
+            """Return sorted node, with children as `(name, node)` list instead of a dictionary."""
+            keys = natsort.natsorted(node[1].keys(), alg=natsort.ns.F | natsort.ns.IC)
+            taxonomy.sort_classifications(keys, lang, level)
+            subnodes = []
+            for key in keys:
+                subnodes.append((key, sort_node(node[1][key], level + 1)))
+            return (node[0], subnodes)
+
+        root = sort_node(root)
+        # Step 3: collapse the tree structure into a linear sorted list,
+        # with a node coming before its children.
+
+        def append_node(classifications, node, path=()):
```

Comment:
May I request a split of all the hierarchical stuff to a special module? `utils.py` is quite cluttered, so some special file for those hierarchies will work better. In that case, sort_node, append_node, and perhaps some others, would also be re-usable between functions. (I kinda dislike defining functions inside functions.)